pkg/router/template/limiter/limiter.go

```diff
@@ -0,0 +1,144 @@
+// only run once when the time allows it.
+func (csrl *CoalescingSerializingRateLimiter) RegisterChange() {
+	glog.V(8).Infof("RegisterChange called")
+
+	csrl.changeWorker(true)
+}
+
+func (csrl *CoalescingSerializingRateLimiter) changeWorker(userChanged bool) {
+	csrl.lock.Lock()
+	defer csrl.lock.Unlock()
+
+	glog.V(8).Infof("changeWorker called")
+
+	if userChanged && csrl.changeReqTime == nil {
+		// They just registered a change manually (and we aren't in the middle of a change)
+		now := time.Now()
+		csrl.changeReqTime = &now
+	}
+
+	if csrl.handlerRunning {
+		// We don't need to do anything else... there's a run in progress, and when it is done it will re-call this function at which point the work will then happen
+		glog.V(8).Infof("The handler was already running (%v) started at %s, returning from the worker", csrl.handlerRunning, csrl.lastStart.String())
+		return
+	}
+
+	if csrl.changeReqTime == nil {
+		// There's no work queued so we have nothing to do.  We should only get here when
+		// the function is re-called after a reload
+		glog.V(8).Infof("No invoke requested time, so there's no queued work.  Nothing to do.")
+		return
+	}
+
+	// There is no handler running, let's see if we should run yet, or schedule a callback
+	now := time.Now()
+	sinceLastRun := now.Sub(csrl.lastStart)
+	untilNextCallback := csrl.callInterval - sinceLastRun
+	glog.V(8).Infof("Checking reload; now: %v, lastStart: %v, sinceLast %v, limit %v, remaining %v", now, csrl.lastStart, sinceLastRun, csrl.callInterval, untilNextCallback)
+
+	if untilNextCallback > 0 {
+		// We want to reload... but can't yet because some window is not satisfied
+		if csrl.callbackTimer == nil {
+			csrl.callbackTimer = time.AfterFunc(untilNextCallback, func() { csrl.changeWorker(false) })
+		} else {
+			// While we are resetting the timer, it should have fired and be stopped.
+			// The first time the worker is called it will know the precise duration
+			// until when a run would be valid and has scheduled a timer for that point
+			csrl.callbackTimer.Reset(untilNextCallback)
+		}
+
+		glog.V(8).Infof("Can't invoke the handler yet, need to delay %s, callback scheduled", untilNextCallback.String())
+
+		return
+	}
+
+	// Otherwise we can reload immediately... let's do it!
+	glog.V(8).Infof("Calling the handler function (for invoke time %v)", csrl.changeReqTime)
+	csrl.handlerRunning = true
+	csrl.changeReqTime = nil
+	csrl.lastStart = now
+
+	// Go run the handler so we don't block the caller
+	go csrl.runHandler()
+
+	return
+}
+
+func (csrl *CoalescingSerializingRateLimiter) runHandler() {
+	// Call the handler, but do it in its own function so we can cleanup in case the handler panics
+	runHandler := func() error {
+		defer func() {
+			csrl.lock.Lock()
+			csrl.handlerRunning = false
+			csrl.lock.Unlock()
+		}()
+
+		return csrl.handlerFunc()
+	}
+	if err := runHandler(); err != nil {
```

Comment:
Nit: The func variable is the same as the function it resides in. Maybe choose a different name here. Spun me around a bit because I thought we are making a recursive call :).